src/protocols/gwmp/hpr_gwmp_redirect_worker.erl

```diff
@@ -0,0 +1,124 @@
+%%%-------------------------------------------------------------------
+%% @doc
+%% == GWMP Redirect Worker ==
+%%
+%% This worker enables remapping source destinations on the fly. Need
+%% to route a packet, but don't know the correct destination until you
+%% know the originating region? This worker fills that need.
+%%
+%% Starts a udp socket with `port'.
+%%
+%% The worker will acknowledge `pull_data' as a courtesy.
+%%
+%% Receiving a `push_data' the worker will pull the `regi' key from
+%% the `stats' map in the payload and return a special message of
+%% where that packet should be sent instead.
+%%
+%% Unmapped regions will be logged and dropped.
+%%
+%% Usage:
+%% /sys.config.src
+%% {hpr, [
+%%     {redirect_by_region, #{
+%%         port => 1777,
+%%         remap => #{
+%%             <<"US915">> => <<"127.0.0.1:1778">>,
+%%             <<"EU868">> => <<"127.0.0.1:1779">>,
+%%             <<"AS923_1">> => <<"127.0.0.1:1780">>
+%%         }
+%%     }}
+%% ]}
+%%
+%% @end
+%%%-------------------------------------------------------------------
+-module(hpr_gwmp_redirect_worker).
+
+-behaviour(gen_server).
+
+-include("semtech_udp.hrl").
+
+%% API
+-export([
+    start_link/1
+]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+-record(state, {
+    port :: non_neg_integer(),
+    remap :: map(),
+    socket :: gen_udp:socket()
+}).
+
+start_link(Args) ->
+    case maps:size(Args) of
+        0 -> ignore;
+        _ -> gen_server:start_link({local, ?MODULE}, ?MODULE, Args, [])
+    end.
+
+init(Args) ->
+    #{port := Port, remap := Map} = Args,
+
+    {ok, Socket} = gen_udp:open(Port, [binary, {active, true}]),
+
+    {ok, #state{
+        port = Port,
+        remap = Map,
+        socket = Socket
+    }}.
+
+handle_call(Request, From, State) ->
+    lager:warning("rcvd unknown call msg: ~p from: ~p", [Request, From]),
+    {reply, ok, State}.
+
+handle_cast(Request, State) ->
+    lager:warning("rcvd unknown cast msg: ~p", [Request]),
+    {noreply, State}.
+
+handle_info(
+    {udp, Socket, Address, Port, IncomingData},
+    #state{socket = Socket, remap = ReMap} = State
+) ->
+    try semtech_udp:identifier(IncomingData) of
+        ?PUSH_DATA ->
+            TxPk = semtech_udp:json_data(IncomingData),
+            Stat = maps:get(<<"stat">>, TxPk),
+            Region = maps:get(<<"regi">>, Stat),
+            case maps:get(Region, ReMap, undefined) of
+                undefined ->
+                    lager:warning([{region, Region}], "dropping redirect for unmapped region"),
+                    dropped;
+                NewDest ->
+                    lager:debug([{region, Region}], "redirecting gateway"),
+                    NewData = jsx:encode(#{
+                        new_dest => NewDest,
+                        packet => erlang:binary_to_list(IncomingData)
```

Comment:
@michaeldjeffrey 
Why is IncomingData converted to a list?

Reply:
jsx tries to treat all binaries as strings when encoding/decoding. 
The `IncomingData` binary is raw data, and most likely doesn't sit wholly in the utf-8 space.

Converting it to a list gives us an array of the code points that jsx will ignore. 
Then we can turn it back into a binary on the other side without losing anything.